### ISSUE TITLE
Enhancing the receiver control and removing the selector control?

### ISSUE THREAD

**Issue by Patowhiz** [africanmathsinitiative/R-Instat#5480]
**Is your feature request related to a problem? Please describe.**
Currently the receiver is used as the ultimate destination of the columns to be used in any purpose that the user desires. The current operation is you select the receiver then you select the column from the selector. This works nice but it also contravenes the principles of a good design and user flow, from left to right. (@lilyclements elaborates this issue in #2868).

**Describe the solution you'd like**
I haven't thought deeply about it's implementation yet but from the design point of view and the much I've gathered in terms of how the selector is currently implemented, I think it can be pushed to work as a popup for the receiver. 

This is how I think this could be designed to, as a user I just see the receivers, when I click on a receiver, a nice popup is shown on top of the receiver, from which I can select the columns I want, or even autocomplete the column name as I type in the receiver. The pop can be designed to have even other controls like selecting the data frame to use, and the data options available. (I have done this kind of UI implementation before, with comment control, where it expand on top of the comment textbox. It's doable with the winforms, and very possible with uwpf which is the future of .Net Core!)

Fundamentally, the popup will be the selector.

This will help reduce the number of controls on the dialogs and maintain principles of good design. With reduction on the number of controls on a dialog, the UX will be greatly improved.

@dannyparsons @rdstern  my above suggestion is based on the current knowledge I have of how the selector relates to the receiver(from the implementation point of view), I still don't know what inspired the design to be the way it is currently, so I might have missed something.

**Comment by dannyparsons** [comment]
I agree that the selector/receiver experience for the user can be improved, and the left/right left/right movements are inefficient. It's good you raise this issue.

One much simpler and less drastic change to improve this (that has been documented for a long time but I can't seem to find the issue now) is to automatically set the focus to the next receiver after filling the previous one. This would probably only be for single receivers (or maybe also if multiple items are added to a multiple receiver) and mean the user doesn't have to leave the selector and go from left to right. To me that's a simpler solution that would need less drastic change.

I think also one of the original reasons for having a left/right selector/receiver layout is that this is standard in most (all?) other menu based statistics software, and so its familiar for users with experience of the common stats packages.

I can see some advantages to a pop up, but my first thoughts are that this would not be as suitable for the kind of users we have. When we have users with little computer literacy who [struggle to use a right click appropriately](https://github.com/africanmathsinitiative/R-Instat/issues/5242#issuecomment-516457997) having pop ups which constantly change what's on the screen, and obscuring other parts of the dialog, I suspect people will find confusing. Closing and opening the pops up might even be a challenge for some. I am thinking of some of the recent uses in climatic workshops.
I think one of the strengths of what we have now is the consistency of the selectors and receiver always being in the same, familiar place on a dialog.
Happy to be disagreed with though! I like that you're thinking of new ways to improve the software.

**Comment by Patowhiz** [comment]
Yeah, I can see how it's important to also think of the end users with little compute literacy, and that has made me rethink the simplicity . Have a look at the design sketch proposal below. We can create a custom combo box like control(We could advance the existing one), and the label can change from just the name to select column: , then the change of the data frames can be pushed just besides it(this will be memorable I think, we could add a tooltip that explains what it does, this small button can open a popup sort of a subdialog from which a user can access data options and select the data frame to use). This will reduce the complexity of many forms both from the user point of view(very important) and even in implementing the code(I think so). 

![rename column](https://user-images.githubusercontent.com/21179007/63945606-48581b80-ca7c-11e9-9777-1c993d9e9622.JPG)

![sketch](https://user-images.githubusercontent.com/21179007/63946660-2069b780-ca7e-11e9-9654-a3a20dae43ab.jpg)

 For multireceivers instead of adding to the listview, we could have a listview with items that are checkboxes. Then a user would just check on the columns that he/she wants 

![multi](https://user-images.githubusercontent.com/21179007/63946656-1f388a80-ca7e-11e9-825c-41109d88cada.JPG)

![sketch 2](https://user-images.githubusercontent.com/21179007/63947109-01b7f080-ca7f-11e9-8455-5e221ece5a01.jpg)

**Comment by rdstern** [comment]
@Patowhiz can we also consider a further issue on the selection of variables - both single and multiple.  We need to be able to handle "wide" data frames better than we do - and that is with whatever "system" for the controls.  An example is the R-package DAAG where the data frame called rockArt has 641 variables.  A recent Kenya survey had 4000 variables!  
SPSS has groups of variables and I assume we could type the first letters and then the choice is restricted?  Currently no typing into the receivers is possible.

**Comment by rdstern** [comment]
And while on the general issue of the dialogues a more crazy suggestion.  On some (long) dialogues and particular screen resolutions I am not able to see the 5 bottom keys.  So I can't press OK.  Could there be a key combination or a possibility to tab to these controls.  My possible key combination would (if possible be 3 keys <ctrl>ok, <ctrl>re, <ctrl>cl, <ctrl> he and <ctrl>sc, but that is very unconventional and some key combinations are tricky to do together.  I'd be happy with just the <ctrl>ok for now!

**Comment by dannyparsons** [comment]
Typing in the receivers and having auto complete are things we thought of from the start but delayed implementing, they would definitely be good improvements.
Also a better display in the multiple receiver, enhancing the list box or otherwise is good to consider.

At the very beginning we considered the combo box type controls you presented but decided against these as they are very difficult for wide data sets as Roger mentioned.
Another improvement related to this is having a "search" in the selector so you can type a column name or pattern for columns which is very useful for wide data sets.

**Comment by Patowhiz** [comment]
@rdstern yes, we could implement the autocomplete in any way that we want(we have partly done this in CLIMSOFT- handling several stations and elements with long names, and the autocomplete shortens the selection process after just typing 3 characters!).

@rdstern I think the proposal to use shortcut keys in such scenarios of hidden controls is a good temporary solution, Sometimes I face the same problems when using my surface computer .  The issue of resolution has generally been a major problem with the winforms. When winforms was designed, at the time, microsoft didn't quite anticipate the future high resolutions on small screens. It's why they later focused on wpf as the new platform. Now that .NET framework is facing it's last days, we are heading towards [.NET Core / .Net 5 ](https://devblogs.microsoft.com/dotnet/introducing-net-5/). Which is coming(by Nov 2020) with good news too, the applications will be able to run on Windows, MacOS and LINUX, saving us tones of work in our future target implementations of R-Instat(Good news for the EU users that requested for R-Instat to be ported to MAC OS). I'm hoping as we transition towards .NET 5 we will embrace [wpf ](https://docs.microsoft.com/en-us/windows/apps/desktop/choose-your-platform) or [uwp ](https://docs.microsoft.com/en-us/windows/apps/desktop/choose-your-platform) as our application platform, to improve on the look and feel of R-Instat.

@dannyparsons by wide dataset, you mean a data set whose field names are long?  We have a feature called [DropDownWidth](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.combobox.dropdownwidth?view=netframework-4.8) which we can manipulate in quite a flexible manner to have the values be visible in combo boxes drop down selections. In Climsoft the Autocomplete feature appeared to do it automatically(as built by microsoft team).

Again most of GUI design changes are hard to determine how users will react to them(until it's tried), the aim is always to make it easier for users, in my experience cluttering the forms with more controls hinds adoption by the not so advanced users, if we aim to reduce the controls but at the same time maintaining the optimum functionality that R-Instant gives then it's a win win. 

I could try to implement 1 simple existing dialog (in consultations with @dannyparsons ) using the proposed single receiver (if agreed it's not a waste of time), then we can determine if it's a good move after the outcome.

**Comment by rdstern** [comment]
@Patowhiz some very interesting, and encouraging ideas here.  I would also hope that @volloholic can join this discussion.
Just one point that the main set of issues I would like to be able to address with wide datasets is not particularly those with ones where the variable names are wide, but datasets where there are many variables, i.e. there are many "columns" of data.  
Where the variable names are big, I would prefer a short name, with the existing name put into the variable label.  
There could be a variety of ways of coping with what I will still call "wide" data sets.  Including:
a) Changing the structure logically into multiple data frames, because "wide" often includes data at multiple levels.
b) Filtering by column as well as row, which we have long had in mind.
c) Defining and using "groups" of variables, which SPSS does quite well.
d) Being able to select variables easily.
e) etc.
We are here often thinking of one of the areas we mentioned originally for R-Instat to support, namely those who process large (traditional) surveys, often government surveys - like the datasets from IPUMS.

**Comment by Patowhiz** [comment]
@rdstern yes, I think such ideas **(a),(b),(c)** will help with the wide data sets, from which we can make the receiver pick the variables from the structured source to enable the user do **(d)**. Again with a good autocomplete I don't think if it will be that hard for users to use the feature. ( [wikipedia ](https://en.wikipedia.org/wiki/Autocomplete) has references to research and studies done on effectiveness of autocomplete ). With current trends of how we interact with applications especially the web, users are used to typing a few characters and using autocomplete feature. They only need to remember a few characters of what they are looking for. 

<img width="156" alt="autocomplete" src="https://user-images.githubusercontent.com/21179007/64065089-cb5eaa80-cc11-11e9-9943-3aa60bcbc351.png">

**Comment by volloholic** [comment]
I am pleased with where this discussion has got to. The autocomplete feature is one we have always agreed needs to be implemented and I would be happy someone got on and did something on this as soon as possible. 
I think the argument for keeping the selector structure has been made but I would like to reinforce the fact that visibly seeing which dataframe you are working in is essential, people already miss this so hiding the data frame further would not be advisable. In most actions R-Instat is still like R working on a single dataframe so this needs to be explicit. Especially since we want to encourage working across dataframes at different levels as implied by a). 
Filtering the columns b) already exists in the form of hiding columns but it is not yet practical or useful. The issue is really comes down to c) how the columns could be sensibly grouped so that they could be hidden in bulk on demand. 
Although autocomplete is a good start for d) it will not be enough on it's own.
Danny's previous point about the auto focus is also really important, when I spend time working with people starting out that is what they struggle with getting the focus in the right place. 
@Patowhiz I am pleased with you wanting to engage with the design aspects. One of the most important things you can do to be constructive with design is to interact with people just starting to use the software. While the theory of good design is useful to be aware of and bring in where possible, human centred design is often more effective at identifying small things that can be done which an ave a big impact.
On the idea of moving to wpf or uwp, that would be an immense amount of work, while it has its benefits it also has its downsides. The biggest part of the work would be to re-imagine the user experience so that we get benefits out of the change. It is something we have considered on a number of occasions and would be essential as part of a long term vision for R-Instat to become multi-platform including mobile! However this is certainly no a top priority at the moment.

@Patowhiz I suggest this issue be closed and that you open issues on the separate actionable items, so that you can get to work actually implementing some of the useful improvements that have come out of this discussion.

**Comment by dannyparsons** [comment]
I'm also not against a more major re-design of dialog but I don't think starting by looking at with this big change is the way to do.

@Patowhiz one thing that already have agreement to help with user experience on dialogs is "Auto switching focus to next receiver after being filled". We definitely have an issue on that somewhere. A suggestion for an implementation of this would be really great. It's not huge work, but will need some details thought out carefully.

Doing this, then possibly followed by implementing typing with autocomplete in the (single) receiver would be big steps. From there I think we would be in a better position to discuss the need for more drastic changes.

**Comment by Patowhiz** [comment]
@dannyparsons I agree going with a sudden big change may not be the best way to go about it. I'll look into enhancing the receiver for the autocomplete option.

I still don't get how the auto switching focus to the next helps this situation. I could be wanting to still change the current selection in the current receiver. With auto switching, the program will be doing against want I want to do and I'll have to always manually select the correct receiver.  I'll find the issue and read more about it.

I also thought we should let this issue open for reference purposes. In future a new proposal could be added into it.

Redesigning of R-Instat is inevitable even though we are still far from it.

**Comment by dannyparsons** [comment]
@Patowhiz see my earlier comment:

> This would probably only be for single receivers (or maybe also if multiple items are added to a multiple receiver)

Think about auto switching focus is for single receivers. After a variable has been put in a single receiver in almost all cases you want the focus to then move to the next receiver. As I said, for a multiple receiver its less clear. Adding a single item to a multiple receiver I think you don't want the focus to change. If you add more than one item in one go, I can see an argument for switching the focus. Or we may decide this only applies to single receivers to keep it simple.

Does that help to clarify?

**Comment by rdstern** [comment]
One obvious dialogue to practice with could be the **_Prepare > Dates > Make Date_**.  With the buttons at the top set to the third option i.e. Year-Month-Day the corresponding single receivers are usually completed in that order.

**Comment by Patowhiz** [comment]
@dannyparsons yes I now get your point. Let me create another issue for the auto switching also.

So in relation to this issue we now have 3 immediate issues: 

- [ ] Enabling typing, autocomplete and selection of variables in single receivers. #5506
- [ ] Display of Groups/Sets of variables in the selector. #4009
- [ ] Enabling auto switching of focus on single receivers on selection of variables in the selectors. #5508

@rdstern was there an issue on the special right click menu that was to go to the receiver? I still don't know what was to go in the context menu.

I've also seen @rdstern had proposed a functional change to the multi-receiver in issue #4069
